src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,919 @@
+;;;; Errors
+(define-constant ERR_PANIC 0)
+(define-constant ERR_NAMESPACE_PREORDER_NOT_FOUND 1001)
+(define-constant ERR_NAMESPACE_PREORDER_EXPIRED 1002)
+(define-constant ERR_NAMESPACE_PREORDER_ALREADY_EXISTS 1003)
+(define-constant ERR_NAMESPACE_UNAVAILABLE 1004)
+(define-constant ERR_NAMESPACE_NOT_FOUND 1005)
+(define-constant ERR_NAMESPACE_ALREADY_EXISTS 1006)
+(define-constant ERR_NAMESPACE_NOT_LAUNCHED 1007)
+(define-constant ERR_NAMESPACE_PRICE_FUNCTION_INVALID 1008)
+(define-constant ERR_NAMESPACE_PREORDER_CLAIMABILITY_EXPIRED 1009)
+(define-constant ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED 1010)
+(define-constant ERR_NAMESPACE_OPERATION_UNAUTHORIZED 1011)
+(define-constant ERR_NAMESPACE_STX_BURNT_INSUFFICIENT 1012)
+(define-constant ERR_NAMESPACE_BLANK 1013)
+(define-constant ERR_NAMESPACE_ALREADY_LAUNCHED 1014)
+(define-constant ERR_NAMESPACE_HASH_MALFORMED 1015)
+(define-constant ERR_NAMESPACE_CHARSET_INVALID 1016)
+
+(define-constant ERR_NAME_PREORDER_NOT_FOUND 2001)
+(define-constant ERR_NAME_PREORDER_EXPIRED 2002)
+(define-constant ERR_NAME_PREORDER_FUNDS_INSUFFICIENT 2003)
+(define-constant ERR_NAME_UNAVAILABLE 2004)
+(define-constant ERR_NAME_OPERATION_UNAUTHORIZED 2006)
+(define-constant ERR_NAME_STX_BURNT_INSUFFICIENT 2007)
+(define-constant ERR_NAME_EXPIRED 2008)
+(define-constant ERR_NAME_GRACE_PERIOD 2009)
+(define-constant ERR_NAME_BLANK 2010)
+(define-constant ERR_NAME_ALREADY_CLAIMED 2011)
+(define-constant ERR_NAME_CLAIMABILITY_EXPIRED 2012)
+(define-constant ERR_NAME_NOT_FOUND 2013)
+(define-constant ERR_NAME_REVOKED 2014)
+(define-constant ERR_NAME_TRANSFER_FAILED 2015)
+(define-constant ERR_NAME_PREORDER_ALREADY_EXISTS 2016)
+(define-constant ERR_NAME_HASH_MALFORMED 2017)
+(define-constant ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH 2018)
+(define-constant ERR_NAME_NOT_RESOLVABLE 2019)
+(define-constant ERR_NAME_COULD_NOT_BE_MINTED 2020)
+(define-constant ERR_NAME_COULD_NOT_BE_TRANSFERED 2021)
+(define-constant ERR_NAME_CHARSET_INVALID 2022)
+
+(define-constant ERR_PRINCIPAL_ALREADY_ASSOCIATED 3001)
+(define-constant ERR_INSUFFICIENT_FUNDS 4001)
+
+(define-constant NAMESPACE_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAMESPACE_LAUNCHABILITY_TTL u10)
+(define-constant NAME_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAME_GRACE_PERIOD_DURATION u5)
+
+(define-constant ATTACHMENTS_INV_PAGE_SIZE u8)
```

Comment:
The point of using pages is to reduce amount of MARF reads required to iterate through the sequence of zone file hashes.  The only way to do this is to ensure that a single MARF read can return multiple zone file hashes -- i.e. multiple zone file hashes are stored as a single "page".  But in order for this to happen, the `attachments-inv` map needs to map a page number to a list of multiple content hashes, not a single hash.

This should be addressed before the code freeze next Tuesday.  I think one of two things should happen:
* We can drop `ATTACHMENTS_INV_PAGE_SIZE` altogether, and remove the `page` element from the key.  The Clarity smart contract would do no pagination at all; it would only track the `index` of each zone file hash.  The Atlas network would employ its own pagination logic and caching to minimize the number of MARF reads (N.B. this can be added after the code freeze; it can be added even after Stacks 2.0 launches).
* We should change the `attachments-inv` to implement true paging, such that one value in this map contains multiple zone file hashes.

I'm in favor of the former since I think it would be less disruptive to the codebase, and it would allow a degree of flexibility in pagination strategies that the BNS contract could not give us (since it can't be updated after it's deployed).  But I'll leave the decision to you.

Reply:
Good point, addressed with a naive variant of your former suggestion - https://github.com/blockstack/stacks-blockchain/pull/2013/commits/2dc4945beffe616add25462b716bf039a3ea8c00